src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,115 @@
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
+    public void growIfArrayIsFull() {
+        if (size == elements.length) {
+            float multipler = 1.5f;
+            int newCapacity = Math.round(elements.length * multipler);
+            System.arraycopy(elements, 0, elements = new Object[newCapacity], 0, size);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayIsFull();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            growIfArrayIsFull();
+            System.arraycopy(elements, index, elements, index + 1, Math.abs(index - size));
+            size++;
+            elements[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            return (T) elements[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            elements[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index
+                    + ", Size: " + size);
+        } else {
+            final Object element = elements[index];
+            System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+            size--;
+            return (T) element;
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            for (int i = 0; i < size; i++) {
+                if (elements[i] == null) {
+                    System.arraycopy(elements, i + 1, elements, i, size - i - 1);
+                    size--;
+                    return element;
+                }
+            }
+        }
+        for (int i = 0; i < size; i++) {
+            if (elements[i] != null && elements[i].equals(element)) {
+                for (int j = i; j < size; j++) {
```

Comment:
Instead of manually shifting elements in the array, use System.arraycopy() to improve performance and readability.